File: backend/app/api/accounts.py
"""
Accounts API endpoints.

All routes are protected by @require_auth. The authenticated user's UUID is
read from g.current_user_id (injected by the decorator) and forwarded to
every service call so data is always scoped to the current user.
"""

from datetime import datetime, timezone
from flask import Blueprint, g, jsonify, make_response, request
from pydantic import ValidationError as PydanticValidationError
from uuid import UUID

from app.schemas.account import (
    AccountCreate,
    AccountUpdate,
    AccountResponse,
)
from app.services import AccountService
from app.utils.auth import require_auth
from app.utils.exceptions import NotFoundError, ValidationError, BusinessRuleError
from app.utils.responses import success_response, error_response
from app.utils.sync_cursor import encode_cursor, decode_cursor

accounts_bp = Blueprint("accounts", __name__, url_prefix="/api/v1/accounts")
account_service = AccountService()


def _parse_client_updated_at(header_value: str | None) -> datetime | None:
    """
    Parse the X-Client-Updated-At header value into a UTC datetime.

    The header must be an ISO-8601 string.  Timezone-naive values are treated
    as UTC.  Returns None when the header is absent or cannot be parsed.

    Args:
        header_value: Raw header string or None

    Returns:
        Parsed datetime (UTC-aware) or None
    """
    if not header_value:
        return None
    try:
        dt = datetime.fromisoformat(header_value)
        if dt.tzinfo is None:
            dt = dt.replace(tzinfo=timezone.utc)
        return dt
    except (ValueError, TypeError):
        return None


@accounts_bp.route("", methods=["GET"])
@require_auth
def list_accounts():
    """
    List all accounts for the authenticated user.

    When the ``If-Sync-Cursor`` request header is present and valid the endpoint
    operates in incremental sync mode: only accounts modified since the cursor
    timestamp are returned as a flat list.  If nothing has changed, 304 is
    returned with no body.  A fresh ``X-Sync-Cursor`` header is always included
    in the response so the client can advance its cursor.

    Balance is intentionally omitted from all responses — it is computed
    locally by the frontend from Dexie transaction data.

    Request Headers:
        Authorization (str): Bearer JWT token.
        If-Sync-Cursor (str, optional): Opaque cursor from a previous response.

    Query Parameters:
        include_archived (bool): Include inactive accounts (default: false).
            Ignored in incremental mode (archived accounts are always included
            so that deletions propagate to the client).

    Returns:
        200: List of accounts (full sync or incremental with changes)
        304: No changes since cursor (incremental mode only)
        401: Authentication required
        500: Internal server error
    """
    try:
        user_id = g.current_user_id
        updated_since = decode_cursor(request.headers.get("If-Sync-Cursor"))
        new_cursor = encode_cursor()

        if updated_since is not None:
            accounts = account_service.get_all(
                user_id=user_id, include_archived=True, updated_since=updated_since
            )
            if not accounts:
                resp = make_response("", 304)
                resp.headers["X-Sync-Cursor"] = new_cursor
                return resp
            data = [
                AccountResponse.model_validate(account).model_dump(mode="json")
                for account in accounts
            ]
            resp = make_response(jsonify({"success": True, "data": data}), 200)
            resp.headers["X-Sync-Cursor"] = new_cursor
            return resp

        include_archived = request.args.get("include_archived", "false").lower() == "true"

        accounts = account_service.get_all(
            user_id=user_id, include_archived=include_archived
        )

        data = [
            AccountResponse.model_validate(account).model_dump(mode="json")
            for account in accounts
        ]

        body, status = success_response(data=data)
        resp = make_response(jsonify(body), status)
        resp.headers["X-Sync-Cursor"] = new_cursor
        return resp

    except Exception as e:
        return error_response(f"Error al listar cuentas: {str(e)}", status_code=500)


@accounts_bp.route("/<uuid:account_id>", methods=["GET"])
@require_auth
def get_account(account_id: UUID):
    """
    Get a single account by ID for the authenticated user.

    Returns 404 if the account does not exist OR belongs to a different user
    (ownership is enforced silently — the record simply isn't found).

    Balance is intentionally omitted — it is computed locally by the frontend.

    Path Parameters:
        account_id (UUID): Account ID

    Returns:
        200: Account details
        401: Authentication required
        404: Account not found
        500: Internal server error
    """
    try:
        account = account_service.get_by_id(account_id, user_id=g.current_user_id)
        data = AccountResponse.model_validate(account).model_dump(mode="json")
        return success_response(data=data)

    except NotFoundError as e:
        return error_response(e.message, status_code=404)
    except Exception as e:
        return error_response(f"Error al obtener cuenta: {str(e)}", status_code=500)


@accounts_bp.route("", methods=["POST"])
@require_auth
def create_account():
    """
    Create a new account for the authenticated user.

    Request Body:
        AccountCreate schema

    Returns:
        201: Created account
        400: Validation error
        401: Authentication required
        500: Internal server error
    """
    try:
        account_data = AccountCreate(**request.json)

        account = account_service.create(
            user_id=g.current_user_id,
            name=account_data.name,
            type=account_data.type.value,
            currency=account_data.currency,
            description=account_data.description,
            tags=account_data.tags,
            client_id=account_data.client_id,
        )

        data = AccountResponse.model_validate(account).model_dump(mode="json")
        return success_response(data=data, message="Cuenta creada exitosamente", status_code=201)

    except PydanticValidationError as e:
        return error_response("Error de validación", status_code=400, errors=e.errors())
    except ValueError as e:
        return error_response(str(e), status_code=400)
    except BusinessRuleError as e:
        return error_response(str(e), status_code=400)
    except NotFoundError as e:
        return error_response(str(e), status_code=404)
    except Exception as e:
        return error_response(f"Error al crear cuenta: {str(e)}", status_code=500)


@accounts_bp.route("/<uuid:account_id>", methods=["PUT"])
@require_auth
def update_account(account_id: UUID):
    """
    Update an existing account for the authenticated user.

    Supports Last-Write-Wins (LWW) conflict detection for offline-first clients.
    When the optional request header X-Client-Updated-At is present its value is
    compared against the record's server-side updated_at timestamp.  If the
    server version is more recent the update is rejected with HTTP 409 and the
    current server state is returned so the client can reconcile.

    Path Parameters:
        account_id (UUID): Account ID

    Request Headers:
        Authorization (str): Bearer JWT token.
        X-Client-Updated-At (str, optional): ISO-8601 timestamp of the version
            the client last observed.  Used for LWW conflict detection.

    Request Body:
        AccountUpdate schema

    Returns:
        200: Updated account
        400: Validation error
        401: Authentication required
        404: Account not found
        409: Conflict — server version is newer than client version
        500: Internal server error
    """
    try:
        account_data = AccountUpdate(**request.json)
        user_id = g.current_user_id

        client_updated_at = _parse_client_updated_at(
            request.headers.get("X-Client-Updated-At")
        )
        if client_updated_at is not None:
            current_account = account_service.get_by_id(account_id, user_id=user_id)
            server_updated_at = current_account.updated_at
            if server_updated_at.tzinfo is None:
                server_updated_at = server_updated_at.replace(tzinfo=timezone.utc)
            if server_updated_at > client_updated_at:
                server_data = AccountResponse.model_validate(current_account).model_dump(
                    mode="json"
                )
                return error_response(
                    "Conflicto: el servidor tiene una version mas reciente de este recurso",
                    status_code=409,
                    errors={"server_version": server_data},
                )

        account = account_service.update(
            account_id=account_id,
            user_id=user_id,
            name=account_data.name,
            type=account_data.type.value if account_data.type else None,
            currency=account_data.currency,
            description=account_data.description,
            tags=account_data.tags,
            active=account_data.active,
        )

        data = AccountResponse.model_validate(account).model_dump(mode="json")
        return success_response(data=data, message="Cuenta actualizada exitosamente")

    except PydanticValidationError as e:
        return error_response("Error de validación", status_code=400, errors=e.errors())
    except NotFoundError as e:
        return error_response(e.message, status_code=404)
    except Exception as e:
        return error_response(f"Error al actualizar cuenta: {str(e)}", status_code=500)


@accounts_bp.route("/<uuid:account_id>", methods=["DELETE"])
@require_auth
def delete_account(account_id: UUID):
    """
    Archive an account (soft delete) for the authenticated user.

    Path Parameters:
        account_id (UUID): Account ID

    Returns:
        200: Account archived successfully
        401: Authentication required
        404: Account not found
        422: Account has transactions/transfers
        500: Internal server error
    """
    try:
        account_service.archive(account_id, user_id=g.current_user_id)
        return success_response(message="Cuenta archivada exitosamente")

    except NotFoundError as e:
        return error_response(e.message, status_code=404)
    except BusinessRuleError as e:
        return error_response(e.message, status_code=422)
    except Exception as e:
        return error_response(f"Error al archivar cuenta: {str(e)}", status_code=500)


@accounts_bp.route("/<uuid:account_id>/permanent", methods=["DELETE"])
@require_auth
def hard_delete_account(account_id: UUID):
    """
    Permanently delete an account (hard delete — irreversible).

    Only succeeds if the account has no transactions or transfers.

    Path Parameters:
        account_id (UUID): Account ID

    Returns:
        200: Account permanently deleted
        401: Authentication required
        404: Account not found
        422: Account has transactions or transfers
        500: Internal server error
    """
    try:
        account_service.delete(account_id, user_id=g.current_user_id)
        return success_response(message="Cuenta eliminada permanentemente")
    except NotFoundError as e:
        return error_response(e.message, status_code=404)
    except BusinessRuleError as e:
        return error_response(e.message, status_code=422)
    except Exception as e:
        return error_response(f"Error al eliminar cuenta: {str(e)}", status_code=500)

File: frontend/src/stores/accounts.ts
/**
 * Accounts Store
 *
 * Why Pinia with Composition API?
 * - Type-safe state management with TypeScript
 * - More intuitive than Vuex (no mutations, simpler actions)
 * - Better DevTools integration
 * - Lightweight and modular
 *
 * This store manages:
 * - Account list and selection
 * - Account balances (computed locally from IndexedDB — never fetched from backend)
 * - CRUD operations for accounts
 *
 * Phase 3 change: write actions (create/update/delete) now follow the
 * offline-first pattern — they write to IndexedDB and the mutation queue
 * immediately, before touching the network. The UI updates without waiting
 * for a server round-trip. The SyncManager (Phase 4) will flush the queue
 * when connectivity is restored.
 *
 * Archive/hard-delete change:
 * - archiveAccount (formerly deleteAccount): sets active=false in IndexedDB
 *   instead of removing the record. The archived account stays in IndexedDB
 *   so it can be restored and so its balance history is preserved locally.
 *   Backend interprets the DELETE verb as a soft-delete (archive).
 * - hardDeleteAccount: permanently removes the record from IndexedDB and
 *   enqueues a delete_permanent mutation so the backend also hard-deletes.
 *   Only allowed when the account has no transactions or transfers.
 * - restoreAccount: flips active back to true and enqueues an update mutation.
 */

import { defineStore } from 'pinia'
import { ref, computed } from 'vue'
import { accountsApi } from '@/api/accounts'
import type { CreateAccountDto, UpdateAccountDto, AccountBalance } from '@/types'
import { db, fetchByIdWithRevalidation, generateTempId, mutationQueue } from '@/offline'
import type { LocalAccount } from '@/offline'

export const useAccountsStore = defineStore('accounts', () => {
  // State
  // Why LocalAccount[] instead of Account[]?
  // LocalAccount extends Account, so all existing consumers of this ref
  // (computed properties, components) continue to work without changes.
  // The extra _sync_* fields are additive.
  const accounts = ref<LocalAccount[]>([])
  const balances = ref<Map<string, AccountBalance>>(new Map())
  const selectedAccountId = ref<string | null>(null)
  const loading = ref(false)
  const error = ref<string | null>(null)

  // Computed
  const activeAccounts = computed(() =>
    accounts.value.filter(account => account.active)
  )

  // Why filter active === false rather than !active?
  // active is always a boolean on LocalAccount (required field on Account).
  // Strict equality guards against undefined on records fetched from older
  // IndexedDB schema versions that pre-date the active column.
  const archivedAccounts = computed(() =>
    accounts.value.filter(account => account.active === false)
  )

  const selectedAccount = computed(() =>
    accounts.value.find(account => account.id === selectedAccountId.value) || null
  )

  const accountsWithBalances = computed(() =>
    accounts.value.map(account => ({
      ...account,
      // Prefer balances.value (updated by adjustBalance on every write and by
      // recomputeBalancesFromTransactions after sync). Falls back to
      // account.balance from IndexedDB (persisted by adjustBalance on writes).
      balance: balances.value.get(account.id)?.balance ?? account.balance ?? 0
    }))
  )

  // ---------------------------------------------------------------------------
  // Helper: normalise balance field from API (may arrive as string)
  // ---------------------------------------------------------------------------
  function normalizeBalance(account: LocalAccount): LocalAccount {
    return {
      ...account,
      balance: account.balance !== undefined ? Number(account.balance) : 0
    }
  }

  // ---------------------------------------------------------------------------
  // Actions — Reads (offline-first, stale-while-revalidate)
  // ---------------------------------------------------------------------------

  async function fetchAccounts() {
    loading.value = true
    error.value = null
    try {
      await refreshFromDB()
    } catch (err: any) {
      error.value = err.message || 'Error al cargar cuentas'
      throw err
    } finally {
      loading.value = false
    }
  }

  async function fetchAccountById(id: string) {
    loading.value = true
    error.value = null
    try {
      const localItem = await fetchByIdWithRevalidation(
        db.accounts,
        id,
        (accountId) => accountsApi.getById(accountId),
        (freshItem) => {
          // Background revalidation callback: update the specific entry.
          // Balance comes from IndexedDB (persisted by adjustBalance) — no
          // backend balance endpoint call needed.
          const normalized = normalizeBalance(freshItem)
          const index = accounts.value.findIndex(a => a.id === id)
          if (index >= 0) {
            // Preserve the locally-tracked balance so the server's list value
            // (which may be stale for pending transactions) doesn't overwrite it.
            const localBalance = balances.value.get(id)?.balance
                               ?? accounts.value.find(a => a.id === id)?.balance
            accounts.value[index] = {
              ...normalized,
              balance: localBalance ?? normalized.balance
            }
          } else {
            accounts.value.push(normalized)
          }
        }
      )

      if (localItem) {
        const normalized = normalizeBalance(localItem)
        const index = accounts.value.findIndex(a => a.id === id)
        if (index >= 0) {
          accounts.value[index] = normalized
        } else {
          accounts.value.push(normalized)
        }
        return normalized
      }
    } catch (err: any) {
      error.value = err.message || 'Error al cargar cuenta'
      throw err
    } finally {
      loading.value = false
    }
  }

  /**
   * Recompute every account's balance from the local transaction and transfer
   * history stored in IndexedDB.
   *
   * Why compute locally instead of calling the /balance API endpoint?
   * The frontend is a standalone app — IndexedDB is the single source of
   * truth for ALL displayed data. Backend endpoints are used only for
   * synchronisation (pushing local mutations, pulling server writes from
   * other devices). The /balance endpoint is unreliable for display because
   * it doesn't include pending offline transactions that haven't synced yet.
   *
   * When to call this:
   *   After wallet:sync-complete — at that point fullReadSync has loaded the
   *   complete transaction history into IndexedDB so the computed total is
   *   authoritative and identical to what the server would return.
   */
  async function recomputeBalancesFromTransactions() {
    const allTransactions = await db.transactions.toArray()
    const allTransfers   = await db.transfers.toArray()

    const computed = new Map<string, number>()

    for (const tx of allTransactions) {
      const current = computed.get(tx.account_id) ?? 0
      const delta = tx.type === 'income' ? Number(tx.amount) : -Number(tx.amount)
      computed.set(tx.account_id, current + delta)
    }

    for (const t of allTransfers) {
      // Source account always loses `amount` (what was sent, in source currency).
      const fromBal = computed.get(t.source_account_id) ?? 0
      computed.set(t.source_account_id, fromBal - Number(t.amount))
      // Destination account gains `destination_amount` for cross-currency transfers,
      // or falls back to `amount` for same-currency transfers where destination_amount
      // is undefined. This mirrors the adjustBalance logic in the transfers store.
      const toBal = computed.get(t.destination_account_id) ?? 0
      computed.set(
        t.destination_account_id,
        toBal + Number(t.destination_amount ?? t.amount)
      )
    }

    for (const [accountId, balance] of computed) {
      const account = accounts.value.find(a => a.id === accountId)
      balances.value.set(accountId, {
        account_id: accountId,
        balance,
        currency: account?.currency ?? 'USD'
      })
      const idx = accounts.value.findIndex(a => a.id === accountId)
      if (idx !== -1) {
        accounts.value[idx] = { ...accounts.value[idx], balance }
      }
      // Persist so the recomputed balance survives the next page reload.
      await db.accounts.update(accountId, { balance })
    }
  }

  // ---------------------------------------------------------------------------
  // Actions — Writes (Phase 3: offline-first pattern)
  //
  // Every write method follows the same three-step sequence:
  //   1. Persist the change to IndexedDB immediately so data survives a reload.
  //   2. Update the reactive ref so the UI reflects the change without waiting
  //      for a network round-trip (optimistic UI).
  //   3. Enqueue a PendingMutation so the SyncManager (Phase 4) can replay
  //      the operation against the server when connectivity is available.
  //
  // The `loading` flag is still set so existing UI spinners continue to work.
  // Because IndexedDB writes are very fast (sub-millisecond in practice),
  // the spinner will barely be visible — but we keep it for consistency with
  // the read actions and to give the UI a hook for any future async work.
  // ---------------------------------------------------------------------------

  async function createAccount(data: CreateAccountDto) {
    // Why generateTempId()?
    // We need a stable local identifier for the new record before the server
    // assigns a real UUID. The 'temp-' prefix lets the SyncManager distinguish
    // locally-created records from server-synced ones at a glance (O(1) check
    // via isTempId()) without an extra database lookup.
    const tempId = generateTempId()
    const now = new Date().toISOString()

    // Build the full local representation of the account.
    // Required fields that the server would normally fill in are given sensible
    // defaults:
    //   - tags: CreateAccountDto.tags is optional, default to empty array.
    //   - active: new accounts are always active.
    //   - balance: 0 until the first transaction is recorded.
    //   - _sync_status: 'pending' signals that this record has an unsent mutation.
    //   - _local_updated_at: used for Last-Write-Wins conflict resolution.
    const localAccount: LocalAccount = {
      id: tempId,
      name: data.name,
      type: data.type,
      currency: data.currency,
      description: data.description,
      tags: data.tags ?? [],
      active: true,
      balance: 0,
      created_at: now,
      updated_at: now,
      _sync_status: 'pending',
      _local_updated_at: now
    }

    loading.value = true
    error.value = null
    try {
      // Step 1 — Write to IndexedDB. This is the source of truth for offline
      // data. If the app closes before sync, the record is preserved here.
      await db.accounts.add(localAccount)

      // Step 2 — Update the reactive ref. Vue re-renders immediately without
      // waiting for the network. The user sees their new account right away.
      accounts.value.push(localAccount)

      // Seed the balances map so accountsWithBalances computed doesn't show
      // undefined for the new account.
      balances.value.set(tempId, {
        account_id: tempId,
        balance: 0,
        currency: data.currency
      })

      // Step 3 — Enqueue the CREATE mutation.
      // client_id in the payload is the same tempId so the backend can use it
      // for idempotency: if the same mutation is replayed due to a network
      // timeout, the server detects the duplicate client_id and returns the
      // already-created account rather than creating a second one.
      await mutationQueue.enqueue({
        entity_type: 'account',
        entity_id: tempId,
        operation: 'create',
        payload: { ...data, client_id: tempId }
      })

      return localAccount
    } catch (err: any) {
      error.value = err.message || 'Error al crear cuenta'
      throw err
    } finally {
      loading.value = false
    }
  }

  async function updateAccount(id: string, data: UpdateAccountDto) {
    const localUpdatedAt = new Date().toISOString()

    loading.value = true
    error.value = null
    try {
      // Step 1 — Patch the IndexedDB record.
      // We use Dexie's update() which does a partial update (like PATCH),
      // only overwriting the fields present in the object literal. The
      // _sync_status and _local_updated_at metadata fields are always included
      // to mark this record as dirty for the sync engine.
      await db.accounts.update(id, {
        ...data,
        _sync_status: 'pending',
        _local_updated_at: localUpdatedAt
      })

      // Step 2 — Update the reactive ref.
      // Object spread preserves all existing fields on the record so we don't
      // accidentally lose fields that are not part of UpdateAccountDto
      // (e.g. created_at, server_id).
      const idx = accounts.value.findIndex(a => a.id === id)
      if (idx !== -1) {
        accounts.value[idx] = {
          ...accounts.value[idx],
          ...data,
          _sync_status: 'pending',
          _local_updated_at: localUpdatedAt
        }
      }

      // Step 3 — Merge optimisation: if there is already a pending CREATE for
      // this entity, collapse the UPDATE into the CREATE payload instead of
      // adding a second queue entry. This reduces the sync to a single POST
      // instead of POST + PATCH.
      const pendingCreate = await mutationQueue.findPendingCreate('account', id)
      if (pendingCreate && pendingCreate.id != null) {
        await mutationQueue.updatePayload(pendingCreate.id, {
          ...pendingCreate.payload,
          ...data
        })
      } else {
        // No pending CREATE — the entity exists on the server. Enqueue a
        // normal UPDATE mutation that the sync engine will send as a PATCH.
        await mutationQueue.enqueue({
          entity_type: 'account',
          entity_id: id,
          operation: 'update',
          payload: data as Record<string, unknown>
        })
      }
    } catch (err: any) {
      error.value = err.message || 'Error al actualizar cuenta'
      throw err
    } finally {
      loading.value = false
    }
  }

  /**
   * Archive an account — soft-delete that keeps the record in IndexedDB with
   * active=false. The account moves from activeAccounts to archivedAccounts
   * in the UI. The backend treats DELETE as a soft-delete (archive).
   *
   * Why keep in IndexedDB instead of removing?
   * Archived accounts must be restorable. They also retain their balance
   * history so net worth calculations remain accurate. Physical removal would
   * lose that history on the next page reload.
   *
   * Special case: if the account was created offline and never synced
   * (pending CREATE in queue), archive is equivalent to cancellation —
   * we remove the record entirely and cancel the CREATE mutation instead
   * of enqueuing a DELETE, because the server never knew about this account.
   */
  async function archiveAccount(id: string) {
    loading.value = true
    error.value = null
    try {
      // Cancellation optimisation: if a pending CREATE exists for this entity
      // the record was never synced to the server. We can remove everything
      // locally without sending any request — the entity never existed server-side.
      const pendingCreate = await mutationQueue.findPendingCreate('account', id)
      if (pendingCreate && pendingCreate.id != null) {
        await mutationQueue.remove(pendingCreate.id)
        await db.accounts.delete(id)
        accounts.value = accounts.value.filter(a => a.id !== id)
        balances.value.delete(id)
        if (selectedAccountId.value === id) {
          selectedAccountId.value = null
        }
        // Early return — no DELETE mutation needed, nothing to send.
        return
      }

      // Step 1 — Update IndexedDB: set active=false so the record is preserved
      // but marked as archived. _sync_status=pending flags it for sync.
      const localUpdatedAt = new Date().toISOString()
      await db.accounts.update(id, {
        active: false,
        _sync_status: 'pending',
        _local_updated_at: localUpdatedAt
      })

      // Step 2 — Update reactive ref: patch the account in-place so the
      // archivedAccounts computed picks it up and activeAccounts drops it.
      const idx = accounts.value.findIndex(a => a.id === id)
      if (idx !== -1) {
        accounts.value[idx] = {
          ...accounts.value[idx],
          active: false,
          _sync_status: 'pending',
          _local_updated_at: localUpdatedAt
        }
      }

      if (selectedAccountId.value === id) {
        selectedAccountId.value = null
      }

      // Step 3 — Enqueue DELETE mutation. The backend interprets DELETE as a
      // soft-delete (archive), so we use the same 'delete' operation that the
      // original deleteAccount used. No behaviour change on the server side.
      await mutationQueue.enqueue({
        entity_type: 'account',
        entity_id: id,
        operation: 'delete',
        payload: { id }
      })
    } catch (err: any) {
      error.value = err.message || 'Error al archivar cuenta'
      throw err
    } finally {
      loading.value = false
    }
  }

  /**
   * Hard-delete an account — permanently removes the record from IndexedDB
   * and enqueues a delete_permanent mutation for the server.
   *
   * Why check transaction/transfer counts before deleting?
   * Hard delete is only allowed when the account has no financial history.
   * The UI disables the button when transactions/transfers exist, but this
   * guard is a belt-and-suspenders check to prevent data corruption if the
   * guard is called programmatically or the UI check is bypassed.
   *
   * Why use delete_permanent operation?
   * The backend distinguishes between soft-delete (DELETE verb → archive) and
   * hard delete (DELETE /accounts/:id/permanent). The 'delete_permanent'
   * operation tells the SyncManager to call the hard-delete endpoint.
   */
  async function hardDeleteAccount(id: string) {
    loading.value = true
    error.value = null
    try {
      // Guard: ensure no transactions reference this account.
      const txCount = await db.transactions.where('account_id').equals(id).count()

      // Guard: ensure no transfers reference this account (either direction).
      // Why two separate counts instead of a combined query?
      // Dexie's .or() on a WhereClause requires both sides to use the same
      // index key. source_account_id and destination_account_id are different
      // fields indexed separately, so two queries and a sum is the idiomatic
      // Dexie approach for an OR across different indexed columns.
      const transferFromCount = await db.transfers.where('source_account_id').equals(id).count()
      const transferToCount = await db.transfers.where('destination_account_id').equals(id).count()

      if (txCount + transferFromCount + transferToCount > 0) {
        throw new Error(
          'Cannot hard-delete an account that has transactions or transfers. Archive it instead.'
        )
      }

      // Step 1 — Remove from IndexedDB entirely.
      await db.accounts.delete(id)

      // Step 2 — Remove from reactive state.
      accounts.value = accounts.value.filter(a => a.id !== id)
      balances.value.delete(id)
      if (selectedAccountId.value === id) {
        selectedAccountId.value = null
      }

      // Step 3 — Enqueue delete_permanent mutation.
      await mutationQueue.enqueue({
        entity_type: 'account',
        entity_id: id,
        operation: 'delete_permanent',
        payload: { id }
      })
    } catch (err: any) {
      error.value = err.message || 'Error al eliminar cuenta permanentemente'
      throw err
    } finally {
      loading.value = false
    }
  }

  /**
   * Restore an archived account — flips active back to true and re-enqueues
   * an update mutation so the backend also restores the account.
   *
   * Why use operation 'update' with { active: true }?
   * The backend restore endpoint is a PATCH with active=true. Using the
   * standard 'update' operation keeps the SyncManager logic simple — it
   * already knows how to send PATCH requests for account updates.
   */
  async function restoreAccount(id: string) {
    loading.value = true
    error.value = null
    try {
      const localUpdatedAt = new Date().toISOString()

      // Step 1 — Update IndexedDB: flip active back to true.
      await db.accounts.update(id, {
        active: true,
        _sync_status: 'pending',
        _local_updated_at: localUpdatedAt
      })

      // Step 2 — Update reactive ref: the account moves from archivedAccounts
      // back to activeAccounts via the computed filters.
      const idx = accounts.value.findIndex(a => a.id === id)
      if (idx !== -1) {
        accounts.value[idx] = {
          ...accounts.value[idx],
          active: true,
          _sync_status: 'pending',
          _local_updated_at: localUpdatedAt
        }
      }

      // Step 3 — Enqueue update mutation with { active: true } payload.
      await mutationQueue.enqueue({
        entity_type: 'account',
        entity_id: id,
        operation: 'update',
        payload: { active: true }
      })
    } catch (err: any) {
      error.value = err.message || 'Error al restaurar cuenta'
      throw err
    } finally {
      loading.value = false
    }
  }

  /**
   * deleteAccount — kept as an alias for archiveAccount to avoid breaking
   * any existing callers that have not yet been migrated to the new name.
   *
   * Why an alias instead of a rename everywhere?
   * Component migrations happen in a later task (FE-4). Keeping this alias
   * means the rename does not break callers in the same PR, and the alias
   * can be removed once all callers have been updated.
   */
  const deleteAccount = archiveAccount

  /**
   * Adjust an account's in-memory balance by a signed delta.
   *
   * Called by the transactions and transfers stores immediately after an
   * offline write so the UI reflects the correct balance without waiting
   * for a sync cycle.
   *
   * Why update both balances.value and accounts.value[idx].balance?
   * accountsWithBalances prefers balances.value (kept in memory, updated by
   * adjustBalance on every write and by recomputeBalancesFromTransactions after sync).
   * accounts.value[idx].balance is the cold-start fallback persisted in IndexedDB.
   */
  function adjustBalance(accountId: string, delta: number) {
    // Derive the current balance from whichever source has it.
    // balances.value is populated by adjustBalance() and recomputeBalancesFromTransactions().
    // accounts.value[idx].balance is the fallback from IndexedDB.
    const current = balances.value.get(accountId)
    const account = accounts.value.find(a => a.id === accountId)
    const currentBalance = current?.balance ?? account?.balance ?? 0
    const newBalance = currentBalance + delta

    // Always write to balances.value — even when there was no prior entry.
    // Without this, a cold-start offline session would only update
    // accounts.value[idx].balance. That update gets wiped by refreshFromDB()
    // (triggered after every sync attempt). balances.value survives
    // refreshFromDB() because it only replaces accounts.value, so
    // accountsWithBalances — which prefers balances.value — stays correct.
    balances.value.set(accountId, {
      account_id: accountId,
      balance: newBalance,
      currency: current?.currency ?? account?.currency ?? 'USD'
    })

    // Also update accounts.value[idx].balance so the fallback path and
    // any consumers that read account.balance directly get the fresh value.
    const idx = accounts.value.findIndex(a => a.id === accountId)
    if (idx !== -1) {
      accounts.value[idx] = { ...accounts.value[idx], balance: newBalance }
    }

    // Persist to IndexedDB (fire-and-forget) so the adjusted balance survives
    // a page reload. Without this, reloading resets balance to 0 from the
    // cached list-endpoint data, which never includes accurate balance.
    db.accounts.update(accountId, { balance: newBalance }).catch(() => {})
  }

  function selectAccount(id: string | null) {
    selectedAccountId.value = id
  }

  function getAccountBalance(accountId: string): number {
    // Prefer balances.value (updated by adjustBalance on every write).
    // Fall back to account.balance from accounts.value (persisted in IndexedDB)
    // so the correct value is returned on page reload.
    const fromMap = balances.value.get(accountId)?.balance
    if (fromMap !== undefined) return fromMap
    return accounts.value.find(a => a.id === accountId)?.balance ?? 0
  }

  /**
   * Re-read accounts from IndexedDB without a background API call.
   * Called after wallet:sync-complete to reflect the SyncManager's fresh data.
   */
  async function refreshFromDB() {
    const data = await db.accounts.toArray()
    accounts.value = data.map(normalizeBalance)
  }

  return {
    // State
    accounts,
    balances,
    selectedAccountId,
    loading,
    error,
    // Computed
    activeAccounts,
    archivedAccounts,
    selectedAccount,
    accountsWithBalances,
    // Actions
    fetchAccounts,
    fetchAccountById,
    recomputeBalancesFromTransactions,
    createAccount,
    updateAccount,
    archiveAccount,
    hardDeleteAccount,
    restoreAccount,
    deleteAccount,
    adjustBalance,
    selectAccount,
    getAccountBalance,
    refreshFromDB
  }
})
